### PR TITLE
Modern CMake configuration refactoring

### PR DESCRIPTION
**Pull request by mloskot**
Minimal build and install features implemented according to [Modern CMake](https://www.youtube.com/watch?v=bsXLMQ6WgIk) best practices for all GEOS targets, libraries and tests.

Require CMake 3.5 or later.
Add `cmake/CMakeSettings.json` sample with convenient defaults for users of [CMake integration in Visual Studio 2017]( https://go.microsoft.com//fwlink//?linkid=834763) (usage: copy to root folder).
Add cmake/geos-configure-file.cmake from Dan Baston (@dbaston).

-----

I tested it with CMake 3.11+ on Linux (GCC, clang) and Windows (Visual Studio).

-----

@robe2, @pramsey, @dbaston  - following our discussions  on IRC, here I pushed complete update of my [ml/modern-cmake](https://github.com/mloskot/geos/tree/ml/modern-cmake) branch, rebased on the latest `master` @ c81850660a323ae736e68e31daa05f8685cc375c

## Tasklist

- [x] Iteration of CMake script clean-ups adding missing `unset`, removing any unnecessary bits and bobs (eg. variables), etc.

- [x] Add GEOS client example configured with CMake as a test to verify `geos-config.cmake` [package configuration file](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html) deployment.

- [x] Update [README.md](https://github.com/mloskot/geos/blob/ml/modern-cmake/README.md) with building, testing, installation and using GEOS with CMake

- [x] Update `.travis.yml` with complete workflow: configure, build, test, install, configure and build client.  Install latest CMake. Remove Autotools jobs.

- [x] Update `.appveyor.yml` with workflow: configure, build, test. Install latest CMake. Remove NMake jobs.

- [ ] Update `.appveyor.yml` with rest of workflow: install, configure and build client. 

- [x] Tag version with git hash, tag or `-dirty` indicator for modified working copy (see [CMakeVersionSource.cmake](https://gitlab.kitware.com/cmake/cmake/blob/master/Source/CMakeVersionSource.cmake) and related scripts) in CMake's codebase)

- [ ] Review compilation flags for supported compilers for both consumers, client and developer

- [ ] Replace ugly old-school `if`-s setting flags per compiler with generator expressions passed to `target_compile_options` and `target_compile_definitions`, for example:

    ```cmake
    $<$<CXX_COMPILER_ID:GNU>:-Wall;-Wextra;-Wpedantic>
    $<$<CXX_COMPILER_ID:MSVC>:NOMINMAX>
    ```

- [ ] Add support for OSX Framework

- [ ] Deal with `tools/CMakeLists.txt` or remove alogn with `geos-config` script

- [ ] Add CPack integration

- [x] Add target for Doxygen build

- [ ] ...

Since I'm not sure if/when I'd be able to complete those TODOs, I'll hand the work over to you now.

### CODE REVIEW

**Comment by mloskot** [comment]
FYI, updated lowering CMake requirement to 3.11, but it may be possible to lower it further eg. to 3.5. I just have no mean to test it against 3.5.

**Comment by strk** [comment]
I tried lowering to 3.10 and got:

CMake Error at tests/xmltester/CMakeLists.txt:15 (add_executable):
  add_executable called with incorrect number of arguments

This is with version 3.10.2 which is what Ubuntu 18.04.1 LTS
ships (supported distribution till April 2020).

**Comment by strk** [comment]
CMake 3.10.2 worked, and subsequent `make` invocation took:

    real    3m36.079s
    user    3m17.767s
    sys     0m18.087s

Subsequent failed tests:

    make: *** No rule to make target 'check'.  Stop.
    make: *** No rule to make target 'dist'.  Stop.
    make: *** No rule to make target 'distcheck'.  Stop.
    make: *** No rule to make target 'uninstall'.  Stop.

The `install` rule worked, but I only see two library files
installed in /usr/local/lib/ with the current time:

    -rw-r--r-- 1 root root 21007490 Oct 29 12:30 /usr/local/lib//libgeos.a
    -rw-r--r-- 1 root root   756596 Oct 29 12:30 /usr/local/lib//libgeos_c.a

No .so files (maybe because they were found already present ?)

**Comment by mloskot** [comment]
> make: *** No rule to make target 'check'. Stop. make: ***

`make test`

> No rule to make target 'dist'. Stop. make: ***
> No rule to make target 'distcheck'. Stop. make: ***
> No rule to make target 'uninstall'. Stop.

"With CPack enabled, "make help" will show all the valid targets"
~https://cmake.org/pipermail/cmake/2014-October/058850.html

That, however, someone will have to work about.

> No .so files (maybe because they were found already present ?)

By default, static libs are generated. Use CMake native [BUILD_SHARED_LIBS](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) to ask for shared.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,342 +1,186 @@
...
+#TODO: option(GEOS_BUILD_OSX_FRAMEWORK "Build shared or static library as Framework Bundle for OSX" OFF)
+#mark_as_advanced(GEOS_BUILD_OSX_FRAMEWORK)
 
-# GEOS release version
-# GEOS C++ library SONAME will use these encoding ABI break at every release
-set(GEOS_VERSION "${GEOS_VERSION_MAJOR}.${GEOS_VERSION_MINOR}.${GEOS_VERSION_PATCH}${GEOS_PATCH_WORD}")
-set(VERSION "${GEOS_VERSION}")
+################################################################################
+# Project settings
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.12)
```
Comment (dbaston) [inline CMakeLists.txt]: Looks like `VERSION_GREATER_EQUAL` was introduced in 3.7.

Comment (mloskot) [reply CMakeLists.txt]: Good catch! I'll fix it.

**Comment by mloskot** [comment]
@dbaston FYI, I've fixed the PR to make it single-commit again, as I prefer it.

## Future of this PR

Since you have expressed interest to expand it, I think it may be a a good idea to consider either

1. Merge it into GEOS master -  CMake is secondary anyway, so users should survive if CMake does not work for a few weeks in the development branch (CI-s jobs can be disabled temporarily)

2. Push it as `modern-cmake` branch in GEOS repo and continue work there, and close this PR.

Either way, I expect it to be easier for you to continue the work since I strongly prefer single-commit pull requests, so my workflow involves `git rebase master` and `git push -f` what obviously will cause trouble in collaborative work.

/cc @strk @robe2 @pramsey

**Comment by dbaston** [comment]
@mloskot I will continue work in a new branch: https://github.com/dbaston/libgeos/commits/modern-cmake

**Comment by mloskot** [comment]
@dbaston Great. Shall we close this PR?

**Comment by pramsey** [comment]
I'm going to close this PR and hope Dan's branch is active.

**Pull request closed without merging**